data/basal.go

```diff
@@ -0,0 +1,166 @@
+	Rate         float64         `json:"rate" bson:"rate" valid:"omitempty,basalrate"`
+	Duration     int             `json:"duration" bson:"duration" valid:"omitempty,basalduration"`
+	Insulin      string          `json:"insulin" bson:"insulin,omitempty" valid:"omitempty,basalinsulin"`
+	Value        int             `json:"value" bson:"value,omitempty" valid:"omitempty,basalvalue"`
+	Suppressed   *SupressedBasal `json:"suppressed" bson:"suppressed,omitempty" valid:"omitempty,required"`
+	Base         `bson:",inline"`
+}
+
+//SupressedBasal represents a suppressed basal portion of a basal
+type SupressedBasal struct {
+	Type         string  `json:"type" bson:"type" valid:"required"`
+	DeliveryType string  `json:"deliveryType" bson:"deliveryType" valid:"basaldeliverytype"`
+	ScheduleName string  `json:"scheduleName" bson:"scheduleName" valid:"omitempty,required"`
+	Rate         float64 `json:"rate" bson:"rate" valid:"omitempty,basalrate"`
+}
+
+const (
+	//BasalName is the given name for the type of a `Basal` datum
+	BasalName = "basal"
+
+	deliveryTypeField = "deliveryType"
+	scheduleNameField = "scheduleName"
+	insulinField      = "insulin"
+	valueField        = "value"
+	rateField         = "rate"
+	durationField     = "duration"
+
+	deliveryTypeTag validate.ValidationTag = "basaldeliverytype"
+	rateTag         validate.ValidationTag = "basalrate"
+	durationTag     validate.ValidationTag = "basalduration"
+	insulinTag      validate.ValidationTag = "basalinsulin"
+	valueTag        validate.ValidationTag = "basalvalue"
+
+	injectedDelivery, scheduledDelivery, suspendDelivery, tempDelivery = "injected", "scheduled", "suspend", "temp"
+
+	levemirInsulin, lantusInsulin = "levemir", "lantus"
+
+	rateValidationLowerLimit     = 0.0
+	durationValidationLowerLimit = 0
+	valueValidationLowerLimit    = 0
+	validationGreaterThanInt     = "Must be greater than %d"
+)
+
+var (
+	allowedDeliveryTypes = map[string]string{injectedDelivery: injectedDelivery, scheduledDelivery: scheduledDelivery, suspendDelivery: suspendDelivery, tempDelivery: tempDelivery}
+	allowedInsulins      = map[string]string{levemirInsulin: levemirInsulin, lantusInsulin: lantusInsulin}
+
+	basalFailureReasons = validate.ErrorReasons{
+		deliveryTypeTag: fmt.Sprintf("Must be one of %s,%s,%s,%s", injectedDelivery, scheduledDelivery, suspendDelivery, tempDelivery),
+		rateTag:         fmt.Sprintf("Must be greater than %f", rateValidationLowerLimit),
+		durationTag:     fmt.Sprintf(validationGreaterThanInt, durationValidationLowerLimit),
+		valueTag:        fmt.Sprintf(validationGreaterThanInt, valueValidationLowerLimit),
+		insulinTag:      fmt.Sprintf("Must be one of %s,%s", levemirInsulin, lantusInsulin),
+	}
+)
+
+//BuildBasal will build a Basal record
+func BuildBasal(obj map[string]interface{}) (*Basal, *Error) {
+
+	base, errs := BuildBase(obj)
+	cast := NewCaster(errs)
+
+	basal := &Basal{
+		DeliveryType: cast.ToString(deliveryTypeField, obj[deliveryTypeField]),
```

Comment:
Hmm, not sure if casting not returning an error (but adding an error and returning an empty string) is going to work well.